pkg/controllers/user/pipeline/controller/pipelineexecution/statesyncer.go

```diff
@@ -0,0 +1,80 @@
+package pipelineexecution
+
+import (
+	"context"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/engine"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/utils"
+	"github.com/rancher/rancher/pkg/ticker"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+const (
+	syncStateInterval = 20 * time.Second
+)
+
+type ExecutionStateSyncer struct {
+	pipelineLister          v3.PipelineLister
+	pipelines               v3.PipelineInterface
+	pipelineExecutionLister v3.PipelineExecutionLister
+	pipelineExecutions      v3.PipelineExecutionInterface
+	cluster                 *config.UserContext
+}
+
+func (s *ExecutionStateSyncer) sync(ctx context.Context, syncInterval time.Duration) {
+	for range ticker.Context(ctx, syncInterval) {
+		logrus.Debugf("Start sync pipeline execution state")
+		s.syncState()
+		logrus.Debugf("Sync pipeline execution state complete")
+	}
+
+}
+
+func (s *ExecutionStateSyncer) syncState() {
+	executions, err := s.pipelineExecutionLister.List("", utils.PipelineInprogressLabel.AsSelector())
+	if err != nil {
+		logrus.Errorf("Error listing PipelineExecutions - %v", err)
+	}
+	if len(executions) < 1 {
+		return
+	}
+	pipelineEngine, err := engine.New(s.cluster)
```

Comment:
maybe we can keep the pipelineEngine as a var of ExecutionStateSyncer, so that we can remove cluster from ExecutionStateSyncer

Reply:
The pipelineEngine needs to be get in sync for now, because Jenkins URL is dug out by search nodes and svc port and it can change in after multiple deploy. Tis can be more elegent when we have proxy to cluster. For now I think I will change the cluster attr to nodelister & serviceLister